### DISCUSSION TITLE
Add support for Windows 1.x/2.x Icon file format

### DISCUSSION THREAD

**Discussion by superbonaci** [ImageMagick/ImageMagick#6670] [Contribute]
### Is your feature request related to a problem? Please describe.

Try to view, edit or convert `.ICO` files from Windows 1.x/2.x.

http://fileformats.archiveteam.org/wiki/Windows_1.0_Icon

Sample file: CARDFILE.ICO opened by ICONEDIT.EXE:

![CARDFILE ICO-SCR](https://github.com/ImageMagick/ImageMagick/assets/8339596/86f4c173-e465-49d1-a306-accca4a02cda)

The data are the black&white pixels, the grey zones are believed to be transparent.

The icon of the program CARDFILE.EXE (CARDFILE.ICO) can be seen in the taskbar, bottom left, when CARDFILE.EXE is minimized:

![CARDFILE ICO-MINIMIZED](https://github.com/ImageMagick/ImageMagick/assets/8339596/5189ff55-892a-4e28-8adf-0b2841566f66)

and sometimes in the About dialog, specially in Microsoft programs:

![CARDFILE-ABOUT](https://github.com/ImageMagick/ImageMagick/assets/8339596/c832b2d9-e849-4d69-bbf2-dff35be75341)

According to this website, 1.x and 2.x share same icons:

https://www.howtogeek.com/733912/a-visual-history-of-windows-icons-from-windows-1-to-11/

Current output:
```
$ convert CARDFILE.ICO CARDFILE.PNG
convert: improper image header `CARDFILE.ICO' @ error/icon.c/ReadICONImage/233.
convert: no images defined `CARDFILE.PNG' @ error/convert.c/ConvertImageCommand/3354.
```

### Describe the solution you'd like

Be able to convert from and to ICO file format (Windows 1.x/2.x). For example from and to PNG.

### Describe alternatives you've considered

GIMP and Pillow don't support it yet:

- https://gitlab.gnome.org/GNOME/gimp/-/issues/9814
- https://github.com/python-pillow/Pillow/issues/7320

### Additional context

Samples:

[CARDFILE.ICO.zip](https://github.com/ImageMagick/ImageMagick/files/12218050/CARDFILE.ICO.zip)

**Comment by superbonaci** [comment]
> Looking at CUSTMADE.ICO, scaling it down from 64x64 to 32x32 would lose information - there are individual pixel differences that you can't show in a 32x32 space - which doesn't sound good to me.

Maybe the sample I provided is a device-independent .ico file, just found out that you can choose to save it as `Device Dependent` or `Device Independent`, so I'm attaching the two so you can see the difference:

![device dependent](https://github.com/python-pillow/Pillow/assets/8339596/96d23c14-7004-4df3-99fc-24f33bddeea5)

[DEV.zip](https://github.com/python-pillow/Pillow/files/12771550/DEV.zip)

**Comment by snibgo** [comment]
Windows 1.0, wow. Is that from the Stone Age?

For the Device-Independent Format, with a filesize of 1038 bytes, containing two images each 64x64 pixels, one bit per pixel, this bash script ignores the header, and extracts the two images into PNG files:
```
{
  head -c 14 > tmp1.bin
  head -c 512 > tmp2.bin
  cat > tmp3.bin
} < CARDFILE.ICO

magick -size 64x64 -depth 1 gray:tmp2.bin out.png
magick -size 64x64 -depth 1 gray:tmp3.bin outb.png
```

This needs a version of `head` that doesn't swallow input bytes beyond number in the `-c` option.

out.png is:

![out](https://github.com/ImageMagick/ImageMagick/assets/33812368/c1fe2198-dfec-4563-8626-768322e8148e)

outb.png is:
![outb](https://github.com/ImageMagick/ImageMagick/assets/33812368/8753c076-ef16-4ebf-8090-ec66b91291f7)

**Comment by superbonaci** [comment]
The Device-Dependent format is 1306 bytes, you can grab all the versions here: https://github.com/python-pillow/Pillow/issues/7320

**Comment by urban-warrior** [comment]
Thanks for the problem report.  We can reproduce the issue and will have a patch to fix it in the GIT main branch @ https://github.com/ImageMagick/ImageMagick later today.  The patch will be available in the beta releases of ImageMagick @ https://imagemagick.org/archive/beta/ by sometime tomorrow.

**Reply by superbonaci** [reply]
Do the beta releases contain binary files (.exe or elf) or is just the source code zipped?

**Comment by superbonaci** [comment]
Thanks a lot!!!!

**Comment by superbonaci** [comment]
For whatever reason it's not working.

**Reply by dlemstra** [reply]
Can you explain what is not working?
